### ISSUE TITLE
[Regression] batch_begin/end callbacks no longer get batch number and size

### ISSUE THREAD

**Issue by Flamefire** [tensorflow/tensorflow#39314]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- TensorFlow installed from (source or binary): Bianry
- TensorFlow version (use command below): 2.2.0

**Describe the current behavior**

The callbacks are documented at https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback#on_train_batch_begin as `logs: dict. Has keys batch and size representing the current batch number and the size of the batch.`

The same used to be the case for `on_train_batch_end` in 2.1 but this seems to have been dropped for unknown reason. In 2.1 the `on_train_batch_begin` had a `logs[`size`]=1` which was plain wrong but the value in `on_train_batch_end` was correct so that could be used.

However in 2.2 the `size` log has been removed from `on_train_batch_end` causing existing code to break.

Furthermore contrary to the documentation for neither callback the size is included

**Describe the expected behavior**

The size should be correctly reported to both callbacks for them to use.

**Standalone code to reproduce the issue**

Instead of coming up with a MWE I quote the code directly:

https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/keras/engine/training.py#L847 does not even pass a `logs` parameter so that is clearly wrong

https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/keras/engine/training.py#L855 uses the result of the `train_function` obtained at https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/keras/engine/training.py#L848 which does not contain batch number or size either.

**Comment by omalleyt12** [comment]
@Flamefire Thanks for the issue! For batch number, you can use the `batch` argument to those methods: `on_train_batch_begin(batch, logs)`

The batch size is no longer passed to the logs, this is expected due a rewrite of `Model.fit`. Could you explain your use case? I can advise on how this can be achieved with the new implementation

**Comment by Flamefire** [comment]
> The batch size is no longer passed to the logs, this is expected due a rewrite of Model.fit

This is surprising as it is a breaking change not mentioned in the release notes and contradicts the above linked documentation and even the current docstring: https://github.com/tensorflow/tensorflow/blob/1186e3f2098793952aa82bf356dfe51b967fb26c/tensorflow/python/keras/callbacks.py#L418

> Could you explain your use case? I can advise on how this can be achieved with the new implementation

I have a callback counting the number of examples processed (doing further statistics later). In general the batch size may not be constant (e.g. trailing batch) and passing the batch size to the callback and only counting batches duplicates the batch size and is hence error prone.

**Comment by omalleyt12** [comment]
> I have a callback counting the number of examples processed (doing further statistics later). In general the batch size may not be constant (e.g. trailing batch) and passing the batch size to the callback and only counting batches duplicates the batch size and is hence error prone.

I think this is best handled by a custom `Metric`:

```
class Counter(tf.keras.metrics.Metric):
  def __init__(self, name=None, dtype=None, **kwargs):
    super(Counter, self).__init__(name, dtype, **kwargs)
    self.count = tf.Variable(0)

  def update_state(self, y_true, y_pred, sample_weight=None):
    batch_size = tf.shape(tf.nest.flatten(y_true))[0]
    self.count.assign_add(batch_size)

  def result(self):
    return tf.identity(self.count)
```

> This is surprising as it is a breaking change not mentioned in the release notes and contradicts the above linked documentation and even the current docstring

Good point, we need to update the docstring

**Comment by Flamefire** [comment]
This looks promising. Can this be used in TF 2.1 too? IIRC the metrics are not passed to the callbacks in 2.1.

2 questions on the code: Why is the `flatten` required? Isn't `y_true` a Tensor with the first dim being the batch size already?   
What is the identity for? It's a no-op isn't it?

**Comment by omalleyt12** [comment]
Yep this should work in TF2.1 as well, here's a full example (had to fix the metric code a bit):

```python
import tensorflow as tf

class Count(tf.keras.metrics.Metric):
  def __init__(self, name=None, dtype=None, **kwargs):
    super(Count, self).__init__(name, dtype, **kwargs)
    self.count = tf.Variable(0)

  def update_state(self, y_true, y_pred, sample_weight=None):
    first_tensor = tf.nest.flatten(y_true)[0]
    batch_size = tf.shape(first_tensor)[0]
    self.count.assign_add(batch_size)

  def result(self):
    return tf.identity(self.count)


class PrintInfo(tf.keras.callbacks.Callback):
  def on_train_batch_end(self, batch, logs):
    print('Batch number: {}'.format(batch))
    print('Samples seen this epoch: {}'.format(logs['counter']))

model = tf.keras.Sequential([tf.keras.layers.Dense(1)])
model.compile(optimizer='sgd', loss='mse', metrics=[Count(name='counter')])
x, y = tf.ones((10, 10)), tf.ones((10, 1))
model.fit(x, y, batch_size=2, callbacks=[PrintInfo()], verbose=2)
```

**Comment by omalleyt12** [comment]
> 2 questions on the code: Why is the flatten required? Isn't y_true a Tensor with the first dim being the batch size already?

It's to handle the case where a multi-output Model is used (just to be robust). If your Model only has 1 output it's not needed. `tf.nest.flatten` will turn any nested Python structure (tuple, list, dict, etc) into a flat list

> What is the identity for? It's a no-op isn't it?

Just to make sure we're not returning the `tf.Variable` directly. It's probably ok to do this, but this protects from accidental modification

**Comment by goldiegadde** [comment]
@Flamefire if you are satisfied with the solution, can you please close this issue ?

**Comment by Flamefire** [comment]
Well it is a breaking change that was not announced so "satisfied" is relative. But there is a solution now.

I'd say once the documentation (e.g. https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback#on_train_batch_begin)  is updated this can be closed as that is part of this report and I'm not sure this has been fixed yet.

**Comment by KTTrev** [comment]
I would also like to use the same technique to use the batch size in `on_train_batch_begin`, so I used the example above by just replacing `on_train_batch_end` with `on_train_batch_begin`, but it doesn't work. Any suggestion to solve this?

**Comment by goldiegadde** [comment]
@Flamefire sorry for not announcing the breaking change, this was a miss. The api docstrings have now been updated in this commit [591fb](https://github.com/tensorflow/tensorflow/commit/3f973428767cfbbbeeb412b3d7e0be8e566591fb). I will close this issue now. Thanks!

@KTTrev for your issue, please file a separate issue with a minimal code sample repro of what you are trying to do and what is not working.

**Comment by Flamefire** [comment]
For anyone else finding this: A better implementation of the Count-metric is likely:

```
class Count(tf.keras.metrics.Metric):
    """Metric which counts the number of examples seen"""
    def __init__(self, name='count', dtype=tf.int64, **kwargs):
        super().__init__(name, dtype, **kwargs)
        self.count = self.add_weight(name, initializer='zeros')

    def update_state(self, y_true, y_pred, sample_weight=None):
        first_tensor = tf.nest.flatten(y_true)[0]
        batch_size = tf.shape(first_tensor)[0]
        self.count.assign_add(tf.cast(batch_size, dtype=self.dtype))

    def result(self):
        return self.count
```

This counts the total number of examples even in Multi-Worker mode and is resettable via `reset_states`
